### PR TITLE
allow to change the application root path

### PR DESCRIPTION
**Pull request by gharlan**
With this PR it is possible to change the application root path used in `TemplateHelper::shorten()`.

### CODE REVIEW

**Comment by JefferyHus** [comment]
```php
dirname(dirname(dirname(dirname(dirname(dirname(__DIR__))))));
```
I believe you can still use something more clear instead of too much dirname calls, like creating a defined path constant in the the `src` directory and get the rootPath you want and easily call that constant inside your class function.

File: src/Whoops/Util/TemplateHelper.php [hunk 1]
```diff
@@ -318,4 +322,24 @@ public function getCloner()
...
+    {
+        $this->applicationRootPath = $applicationRootPath;
+    }
+
+    /**
+     * Return the application root path.
+     *
+     * @return string
+     */
+    public function getApplicationRootPath()
```
Comment (staabm) [inline src/Whoops/Util/TemplateHelper.php]: We dont need this getter?

File: src/Whoops/Handler/PrettyPageHandler.php [hunk 2]
```diff
@@ -144,6 +149,12 @@ public function handle()
             $helper->setCloner($cloner);
         }
 
+        $rootPath = $this->applicationRootPath;
```
Comment (staabm) [inline src/Whoops/Handler/PrettyPageHandler.php]: Cant we just do the init in __construct()?

Comment (staabm) [inline src/Whoops/Handler/PrettyPageHandler.php]: .. in one line to define the default..

**Comment by staabm** [comment]
This change is required for cases where composer dependencies are not stored in /vendor/* but a deeper nested subdirectory

**Comment by staabm** [comment]
@JefferHus this PR adds a new feature and I dont think we should change the default right now.. it doesnt look that pretty but did work in the past very well

File: src/Whoops/Util/TemplateHelper.php [hunk 3]
```diff
@@ -106,9 +111,8 @@ public function breakOnDelimiter($delimiter, $s)
      */
     public function shorten($path)
     {
-        $dirname = dirname(dirname(dirname(dirname(dirname(dirname(__DIR__))))));
-        if ($dirname != "/") {
-            $path = str_replace($dirname, '&hellip;', $path);
+        if ($this->applicationRootPath != "/") {
```
Comment (staabm) [inline src/Whoops/Util/TemplateHelper.php]: I wouldnt define a default of '/' here .. take the root path in account when defined and give it an empty/null default. The prettypagehandler should be in charge to provide one

File: src/Whoops/Handler/PrettyPageHandler.php [hunk 4]
```diff
@@ -144,6 +149,12 @@ public function handle()
...
+        if (!$rootPath) {
+            $rootPath = dirname(dirname(dirname(dirname(dirname(dirname(__DIR__))))));
```
Comment (sergeyklay) [inline src/Whoops/Handler/PrettyPageHandler.php]: This will work only for installation via Composer to the project's vendor dir? right?

Comment (staabm) [inline src/Whoops/Handler/PrettyPageHandler.php]: Thats what this PR is about. Before it only worked for the projects vendor dir. With the new api you can instruct the prettypagehandler with whatever dir/path the project is using for the vendor dependencies

File: src/Whoops/Util/TemplateHelper.php (hunk 1 above)
Comment (gharlan) [inline src/Whoops/Util/TemplateHelper.php]: All setters in this class have getters, too. That's the reason why I've added the getter.

File: src/Whoops/Handler/PrettyPageHandler.php [hunk 5]
```diff
@@ -98,6 +103,9 @@ public function __construct()
 
         // Add the default, local resource search path:
         $this->searchPaths[] = __DIR__ . "/../Resources";
+
+        // root path for ordinary composer projects
+        $this->applicationRootPath = dirname(dirname(dirname(dirname(dirname(dirname(__DIR__))))));;
```
Comment (staabm) [inline src/Whoops/Handler/PrettyPageHandler.php]: 2x `;`

**Comment by gharlan** [comment]
Should be fine now.

File: src/Whoops/Handler/PrettyPageHandler.php (hunk 5 above)
Comment (gharlan) [inline src/Whoops/Handler/PrettyPageHandler.php]: done

**Comment by staabm** [comment]
@denis-sokolov your opinion on this?

File: src/Whoops/Handler/PrettyPageHandler.php [hunk 6]
```diff
@@ -98,6 +103,9 @@ public function __construct()
 
         // Add the default, local resource search path:
         $this->searchPaths[] = __DIR__ . "/../Resources";
+
+        // root path for ordinary composer projects
+        $this->applicationRootPath = dirname(dirname(dirname(dirname(dirname(dirname(__DIR__))))));
```
Comment (denis-sokolov) [inline src/Whoops/Handler/PrettyPageHandler.php]: Whoever uses `TemplateHelper::shorten` wants this default to work, so let’s move this line into the constructor of `TemplateHelper`.

**Comment by denis-sokolov** [comment]
@staabm, thanks for doing your review, I appreciate the work you save me!
@gharlan, thanks for this PR! As soon as we address my requested changes, I will merge.

**Comment by gharlan** [comment]
@denis-sokolov should be done now.

**Comment by staabm** [comment]
👍

**Comment by denis-sokolov** [comment]
Beautiful, thank you!

**Comment by staabm** [comment]
@denis-sokolov thx. any plans for a new release-tag?

**Comment by denis-sokolov** [comment]
There you go: 2.1.6

**Comment by staabm** [comment]
thx awesome!

**Pull request merged**